### DISCUSSION TITLE
ESP32 Boost Microphone (INMP441) with AudioEffects

### DISCUSSION THREAD

**Discussion by krabbs** [pschatzmann/arduino-audio-tools#153] [Q&A] [answered]
Hi,

i have connected an INMP441 with an I2SStream to my ESP32 huzzah feather. The example code is running fine.
However i want to play with AudioEffect like to Boost the volume. 

Although I can successfully add effecte to different sources, I fail with a "GeneratorFromStream" source. 
My source code is running on the ESP32 and a signal is also received and output. But the output sound is heavily distorted and is not recognizable anymore I suspect that I have a schematic error and need help with my troubleshooting.

My source code is pretty much the basic setup:
```

#include <Arduino.h>
#include "AudioTools.h"
using namespace audio_tools;  

AudioWAVServer server("XXX","XXX"); 

//Audio
I2SStream i2sStream;    // Access I2S as stream
ConverterFillLeftAndRight<int16_t> filler(LeftIsEmpty); // fill both channels - or change to RightIsEmpty
AudioEffects<GeneratorFromStream<int16_t>> effects(i2sStream);
GeneratedSoundStream<int16_t> in(effects); 


// Audio Format
const int sample_rate = 10000;
const int channels = 1;
float volumeControl = 2.0;


void setup(){
  pinMode(13, OUTPUT);    // RED ON BOARD LED
  digitalWrite(13, HIGH);
  Serial.begin(115200);
  AudioLogger::instance().begin(Serial, AudioLogger::Warning); //Warning
  // setup effects
  effects.addEffect(new Boost(volumeControl));  

  // start i2s input with default configuration
  Serial.println("starting I2S...");
  auto config = i2sStream.defaultConfig(RX_MODE);
  config.i2s_format = I2S_STD_FORMAT; // if quality is bad change to I2S_LSB_FORMAT https://github.com/pschatzmann/arduino-audio-tools/issues/23
  config.bits_per_sample = 16;
  config.channels = channels;
  config.sample_rate = sample_rate;

  i2sStream.begin(config);

  Serial.println("I2S started");

  // start data sink
  server.begin(in, config, &filler);
  in.begin(config);

  digitalWrite(13, LOW);
}

// Arduino loop  
void loop() {
  server.doLoop();  
}

```

**Comment by pschatzmann** [comment, marked as answer]
I think the key question is the following: is it also distorted if the factor is smaller then 1.0 ?
If you just want to control the volume I suggest to use the [VolumeControl](https://github.com/pschatzmann/arduino-audio-tools/wiki/Volume-Control) instead.

**Reply by krabbs** [reply]
it is even disorted when i comment out the line:
`  effects.addEffect(new Boost(volumeControl));  `

This is the reasion, i think there is some faults on my side in the first few lines (maybe because incompatible audio source):

```
//Audio
I2SStream i2sStream;    // Access I2S as stream
ConverterFillLeftAndRight<int16_t> filler(LeftIsEmpty); // fill both channels - or change to RightIsEmpty
AudioEffects<GeneratorFromStream<int16_t>> effects(i2sStream);
GeneratedSoundStream<int16_t> in(effects); 

```

**Comment by pschatzmann** [comment]
I suggest to Remove ConverterFillLeftAndRight<int16_t> filler(LeftIsEmpty); because that's only relevant for 2 channels...

**Reply by krabbs** [reply]
no success.

With the VolumeControl it kind of works.
I will investigate my main problem further 

Thanks for helping
